model/plugin_kvset_options.go

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) 2017-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package model
+
+import (
+	"encoding/json"
+	"net/http"
+)
+
+// PluginKVSetOptions contains information on how to store a value in the plugin KV store.
+type PluginKVSetOptions struct {
+	EncodeJSON      bool        // If true, store the JSON encoding of newValue
+	Atomic          bool        // Only store the value if the current value matches the oldValue
+	OldValue        interface{} // The value to compare with the current value when Atomic is true
+	ExpireInSeconds int64       // Set an expire counter
+}
+
+// IsValid return nil if the chosen options are valid.
+func (opt *PluginKVSetOptions) IsValid() *AppError {
+	if !opt.Atomic && opt.OldValue != nil {
+		return NewAppError(
+			"PluginKVSetOptions.IsValid",
+			"model.plugin_kvset_options.is_valid.old_value.app_error",
+			nil,
+			"",
+			http.StatusBadRequest,
+		)
+	}
+
+	return nil
+}
+
+// GetPluginKeyValue return a PluginKeyValue given a pluginID and a KV pair.
+func (opt *PluginKVSetOptions) GetPluginKeyValue(pluginId, key string, value interface{}) (*PluginKeyValue, *AppError) {
```

Comment:
It feels a bit weird to have this be on `PluginKVSetOptions` instead of having a `NewPluginKeyValueFromOptions(pluginId, key, value, options)`. Thoughts?

Reply:
:+1: